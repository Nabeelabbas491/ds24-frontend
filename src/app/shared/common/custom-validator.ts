import { FormArray, ValidatorFn, AbstractControl } from '@angular/forms';

export class CustomValidators {
  static validateTimeRange(): ValidatorFn {
    return (control: AbstractControl): { [key: string]: any } | null => {
      const timeRangeArray = control as FormArray;
      const timeRanges: [number, number][] = timeRangeArray.controls.map((control: any) => {
        const { from, to } = control.value;
        return [this.convertToNumber(from), this.convertToNumber(to)];
      });

      for (let index = 0; index < timeRanges.length; index++) {
        const [currentStartTime, currentEndTime] = timeRanges[index];
        if (currentStartTime > currentEndTime) {
          timeRangeArray.controls[index].setValidators([this.isStartMore()]);
        } else {
          const overlaps = timeRanges.some(([startTime, endTime], listIndex) => {
            return listIndex !== index && this.isTimeOverlapping(currentStartTime, currentEndTime, startTime, endTime);
          });

          if (overlaps) {
            timeRangeArray.controls[index].setErrors({ overlap: true });
          } else {
            timeRangeArray.controls[index].setErrors(null);
          }
        }
      }

      if (timeRangeArray.controls.some(control => control.invalid)) {
        return { hasErrors: true };
      }

      return null;
    };
  }

  static isStartMore(): ValidatorFn {
    return (control: AbstractControl): { [key: string]: any } | null => {
      const fromControl = control.get('from');
      const toControl = control.get('to');

      if (!fromControl || !toControl) {
        return null;
      }

      const start = this.convertToNumber(fromControl.value);
      const end = this.convertToNumber(toControl.value);

      if (start >= end) {
        return { isStartMore: true };
      }

      return null;
    };
  }

  public static convertToNumber(timeString: any): number {
    return Number(timeString.replace(':', ''));
  }

  public static isTimeOverlapping(
    enteredStartTime: number,
    enteredEndTime: number,
    startTime: number,
    endTime: number,
  ): boolean {
    return (
      (enteredStartTime > startTime && enteredStartTime < endTime) ||
      (startTime > enteredStartTime && startTime < enteredEndTime) ||
      (enteredStartTime === startTime && enteredEndTime === endTime)
    );
  }
}
