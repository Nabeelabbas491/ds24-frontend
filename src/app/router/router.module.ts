import { NgModule } from '@angular/core';
import { Routes, RouterModule } from '@angular/router';
import { accessTokenResolver } from './access-token-resolver';

export const routes: Routes = [
  { path: '', redirectTo: '/vendor', pathMatch: 'full' },
  {
    path: 'booking-template-module',
    loadChildren: () => import('./../screens/booking-template/booking-template.module').then(m => m.BookingTempModule),
    resolve: [accessTokenResolver],
  },
  {
    path: 'vendor',
    loadChildren: () => import('../screens/vendor/vendor.module').then(m => m.VendorModule),
    resolve: [accessTokenResolver],
  },
  {
    path: 'calendar',
    loadChildren: () => import('../screens/calendar/calendar.module').then(m => m.CalendarScreenModule),
  },
  {
    path: 'google-calendar',
    data: { title: 'Google Calendar' },
    loadChildren: () => import('../screens/google-calendar/google-calendar.module').then(m => m.GoogleCalendarModule),
  },
  {
    path: '**',
    loadChildren: () => import('../screens/not-found/not-found.module').then(m => m.NotFoundScreenModule),
    resolve: [accessTokenResolver],
  },
];

@NgModule({
  imports: [RouterModule.forRoot(routes)],
  exports: [RouterModule],
})
export class AppRouterModule {}
