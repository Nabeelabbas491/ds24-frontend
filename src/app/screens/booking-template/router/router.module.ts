import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';
import { BookingTemplateComponent } from '../components/booking-template/booking-template.component';

const routes: Routes = [
  {
    path: '',
    component: BookingTemplateComponent,
  },
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule],
})
export class BookingTempRoutingModule {}
