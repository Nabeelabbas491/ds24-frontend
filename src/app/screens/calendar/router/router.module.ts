import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';
import { CalendarPageComponent } from '../components/calendar-page.component';
import { bookingTemplateResolver } from './booking-template.resolver';

export const routes: Routes = [
  {
    path: ':pid',
    component: CalendarPageComponent,
    data: { title: 'Calendar', page: 'calendar' },
    resolve: { bookingTemplateResolver },
  },
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule],
})
export class CalendarRoutingModule {}
