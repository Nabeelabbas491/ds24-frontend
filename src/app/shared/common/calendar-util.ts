import {
  addDays,
  addMinutes,
  addMonths,
  differenceInDays,
  differenceInMinutes,
  eachDayOfInterval,
  format,
  getDaysInMonth,
  isMonday,
  isSunday,
  nextSunday,
  parseISO,
  previousMonday,
} from 'date-fns';
import { Slot } from './../../types/calendar-day.type';
import { Availability } from './../../types/availability';
import { CalendarMode } from './../../types/misc.type';
import { BookingTemplateDetail } from './../../types/booking.type';
import { Day } from './../../types/calendar-day.type';
import { dayFactory } from './util';
import { Availibility } from 'src/app/types/vendor.types';

export function getInitialCurrentMonth() {
  const date = new Date();
  date.setDate(1);

  return date.toJSON();
}

export function getDaysForCalendar(
  month: string,
  selectedDay: Day | null,
  bookingTemplateDetail: BookingTemplateDetail,
): Day[] {
  const daysArr: Day[] = [];

  const daysInMonth = getDaysInMonth(parseISO(month));

  for (let index = 0; index < daysInMonth; index++) {
    const dayString: string = format(addDays(parseISO(month), index), 'yyyy-MM-dd');
    const slots: Slot[] = [];
    const day: Day = dayFactory([{ dayTitle: dayString, slotsArr: slots }], dayString);

    day.isSelectable = isDayAvailable(dayString, bookingTemplateDetail.availabilities);

    if (selectedDay) {
      day.isSelected = day.dayMonthYearTitle === selectedDay?.dayMonthYearTitle;
    }

    daysArr.push(day);
  }

  return adjustDaysForCalendar(daysArr);
}
export function getDaysForOverrides(month: string, selectedDays: Day[] | null) {
  const daysArr: Day[] = [];
  const daysInMonth = getDaysInMonth(parseISO(month));

  for (let index = 0; index < daysInMonth; index++) {
    const dayString: string = format(addDays(parseISO(month), index), 'yyyy-MM-dd');

    const day: Day = dayFactory([{ dayTitle: dayString, slotsArr: [] }], dayString);
    if (selectedDays && selectedDays.length > 0) {
      day.isSelected = selectedDays.find(d => d.dayMonthYearTitle === day.dayMonthYearTitle) !== undefined;
    }
    daysArr.push(day);
  }

  return adjustDaysForCalendar(daysArr);
}

export function adjustDaysForCalendar(days: Day[]): Day[] {
  const firstDay: Date = parseISO(days[0].dayMonthYearTitle);
  const lastDay: Date = parseISO(days[days.length - 1].dayMonthYearTitle);

  days.unshift(...adjustStartDays(firstDay));
  days.push(...adjustEndDays(lastDay));

  return days;
}

export function adjustStartDays(firstDay: Date): Day[] {
  const daysArr: Day[] = [];

  if (!isMonday(firstDay)) {
    const numberOfDays = differenceInDays(firstDay, previousMonday(firstDay));
    for (let index = 0; index < numberOfDays; index++) {
      const dayString: string = format(addDays(firstDay, -(index + 1)), 'yyyy-MM-dd');
      daysArr.unshift(dayFactory([{ dayTitle: dayString, slotsArr: [] }], dayString, true));
    }
  }
  return daysArr;
}
export function adjustEndDays(lastDay: Date): Day[] {
  const daysArr: Day[] = [];

  if (!isSunday(lastDay)) {
    const noOfDays = differenceInDays(nextSunday(lastDay), lastDay);

    for (let index = 0; index < noOfDays; index++) {
      const dayString: string = format(addDays(lastDay, index + 1), 'yyyy-MM-dd');
      daysArr.push(dayFactory([{ dayTitle: dayString, slotsArr: [] }], dayString, true));
    }
  }
  return daysArr;
}

export function isDayAvailable(currentDate: string, availability: Availability[]): boolean {
  const dayName = format(parseISO(currentDate), 'EEEE').toLowerCase();

  const dateOverride = availability?.find(
    x =>
      x.type === 'override' &&
      x.startDate &&
      x.endDate &&
      isDateWithinDateRange(x.startDate, x.endDate, currentDate) &&
      x.isAvailable,
  );

  if (dateOverride) {
    return true;
  }

  const dayAvailability = availability?.find(x => x.type === 'default' && x.day === dayName && x.isAvailable);
  if (dayAvailability) {
    return true;
  }

  return false;
}

export function isMonthWithinDateRange(startDate: string, endDate: string, currentMonth: string): boolean {
  const startDateFns = parseISO(startDate);
  const endDateFns = parseISO(endDate);
  const currentMonthFns = parseISO(currentMonth);
  const nextMonthFns = addMonths(currentMonthFns, 1);

  if (startDateFns > endDateFns) {
    return false;
  }

  if (startDateFns <= currentMonthFns && endDateFns >= currentMonthFns) {
    return true;
  }

  if (startDateFns >= currentMonthFns && startDateFns < nextMonthFns) {
    return true;
  }

  return false;
}

export function isDateWithinDateRange(startDate: string, endDate: string, currentDate: string): boolean {
  const startDateFns = parseISO(startDate);
  const endDateFns = parseISO(endDate);
  const currentDateFns = parseISO(currentDate);

  if (startDateFns > endDateFns) {
    return false;
  }

  if (startDateFns <= currentDateFns && endDateFns >= currentDateFns) {
    return true;
  }

  return false;
}
export function getTimeSlices(from: string, to: string, duration: number): Slot[] {
  const startDate = parseISO(`${format(new Date(), 'yyyy-MM-dd')} ${from}`);

  const endDate = parseISO(`${format(new Date(), 'yyyy-MM-dd')} ${to}`);

  const minutesDuration = differenceInMinutes(endDate, startDate);

  const numberOfSlices = Math.floor(minutesDuration / duration);

  if (numberOfSlices === 0) {
    return [];
  }

  const slicesArr: Slot[] = [];
  for (let i = 0; i < numberOfSlices; i++) {
    slicesArr.push({
      startTime: format(addMinutes(startDate, duration * i), 'HH:mm'),
      endTime: format(addMinutes(startDate, duration * i + duration), 'HH:mm'),
    });
  }
  return slicesArr;
}
export function checkIfDaySelected(mode: CalendarMode, monthDay: Day, currentDay: Day | null): boolean {
  let isSelected: boolean = false;
  if (mode === 'clientbooking') {
    isSelected = currentDay?.dayMonthYearTitle === monthDay.dayMonthYearTitle;
  } else {
    if (currentDay?.dayMonthYearTitle === monthDay.dayMonthYearTitle) {
      isSelected = !monthDay.isSelected;
    } else {
      isSelected = monthDay.isSelected;
    }
  }
  return isSelected;
}

export function selectCalendarDay(days: Day[], selectedDay: Day): Day[] {
  //select or unselect a day when calendar is set to allow multiple days selection
  const dayAlreadySelected: Day | undefined = days.find(d => d.number === selectedDay.number);

  if (dayAlreadySelected) {
    //if day was already selected, then unselect it
    return days.filter(d => d.number !== selectedDay.number);
  } else {
    days.push(selectedDay);
    return days;
  }
}

export function checkIfDayRangeIncludesSelectedDay(day: Availibility, selectedDay: Day | null): boolean {
  const startDateDate = new Date(day.startDate as string);
  const endDateDate = new Date(day.endDate as string);

  const dateArray = eachDayOfInterval({ start: startDateDate, end: endDateDate });
  const formattedDateArray = dateArray.map(date => format(date, 'yyyy-MM-dd'));

  const selectedDayTitle = selectedDay?.dayMonthYearTitle as string;
  return formattedDateArray.includes(selectedDayTitle);
}
