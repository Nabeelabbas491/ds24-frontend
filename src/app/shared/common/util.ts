import { AbstractControl, ValidationErrors } from '@angular/forms';
import { BreakPoint, LocalStorageKey, PageBase } from 'src/app/types/misc.type';
import { CalendarEntry } from '../../types/calendar-entry.type';
import { Day } from './../../types/calendar-day.type';
import { getDate, parseISO, isToday, isPast, format, parse, isSameMonth } from 'date-fns';
import { BreakpointState, Breakpoints } from '@angular/cdk/layout';
import { AppointmentDetails, BookingTemplate, EventData } from 'src/app/types/vendor.types';
import * as helper from './util';
import { HttpErrorResponse } from '@angular/common/http';
import { MAPPED_TIMEZONES } from './timezones';
import { Tab } from '../../types/tab.type';
import { SaveVendorSetting } from './../../types/vendor-settings.type';
import { Availability } from './../../types/availability';
import { GoogleIntegrationInfo } from 'src/app/types/google-integration.type';
import { BookingSnapshot } from './../../types/booking.type';

export const DEFAULT_PRODUCT_IMAGAE = './assets/images/product_thumb.png';
export const TEMP_VENDOR_NAME = 'James luke'; //this is as per figma design, once we start getting the vendor name from the API, we'll remove this
export const NO_AUTHORIZATION_TOKEN = 'NO_AUTHORIZATION_TOKEN';
export const GOOGLE_CONNECTED_MESSAGE = 'You are already Connected to Google Calendar.';

export const NO_AUTHORIZATION_HEADER = { headers: { NO_AUTHORIZATION_TOKEN: 'true' } };

export const dayFactory = (
  calendarEntries: CalendarEntry[],
  dateString: string,
  isOtherMonth: boolean = false,
): Day => {
  const day: Day = {
    number: getDate(parseISO(dateString)),
    isToday: isToday(parseISO(dateString)),
    isSelected: false,
    isOtherMonth: isOtherMonth,
    isPast: !isToday(parseISO(dateString)) && isPast(parseISO(dateString)),
    dayMonthYearTitle: dateString,
    isSelectable: true,
  };

  const calendarEntry = calendarEntries.find(x => x.dayTitle === day.dayMonthYearTitle);
  if (calendarEntry) {
    day.slotsArr = [...calendarEntry.slotsArr];
  }
  return day;
};

export const convertToNumber = (timeString: string) => {
  return Number(timeString.replace(':', ''));
};

export const getBreakPoint = (breakpointState: BreakpointState): BreakPoint => {
  if (breakpointState.breakpoints[Breakpoints.XSmall]) {
    return 'xsmall';
  }
  if (breakpointState.breakpoints[Breakpoints.Small]) {
    return 'small';
  }
  if (breakpointState.breakpoints[Breakpoints.Medium]) {
    return 'medium';
  }
  if (breakpointState.breakpoints[Breakpoints.Large]) {
    return 'large';
  }
  if (breakpointState.breakpoints[Breakpoints.XLarge]) {
    return 'xlarge';
  }
  return 'undefined';
};
export const bookingDurations = [
  { text: 'DEFAULT_BOOKING_TIME.FIFTEEN_MINUTES', value: '15' },
  { text: 'DEFAULT_BOOKING_TIME.THIRTY_MINUTES', value: '30' },
  { text: 'DEFAULT_BOOKING_TIME.FORTY_FIVE_MINUTES', value: '45' },
  { text: 'DEFAULT_BOOKING_TIME.SIXTY_MINUTES', value: '60' },
  { text: 'DEFAULT_BOOKING_TIME.NINETY_MINUTES', value: '90' },
  { text: 'DEFAULT_BOOKING_TIME.ONE_HUNDRED_TWENTY_MINUTES', value: '120' },
];

export function checkBookingType(control: AbstractControl) {
  const outbound = control?.get('outbound')?.value;
  const inbound = control?.get('inbound')?.value;
  const video = control?.get('video')?.value;

  if (!outbound && !inbound && !video) {
    return { atLeastOneSelected: true };
  }

  return null;
}

export function updateBookingTemplates(
  editedBookingTemplate: BookingTemplate | undefined,
  updatedBookingTemplates: BookingTemplate[],
) {
  const allUpdatedBookingTemplates = editedBookingTemplate
    ? updatedBookingTemplates.map(template => {
        if (template.id === editedBookingTemplate.id) {
          return { ...template, ...editedBookingTemplate };
        }
        return template;
      })
    : updatedBookingTemplates;

  return allUpdatedBookingTemplates;
}
export function guessTimeZone() {
  const timeZone = helper.getClientTimeZone();

  if (timeZone) {
    return timeZone;
  } else {
    return 'Europe/Berlin';
  }
}
export function getClientTimeZone(): string {
  return Intl.DateTimeFormat().resolvedOptions().timeZone;
}

export function returnFormattedDay(selectedDay: Day): string {
  return format(parseISO(selectedDay.dayMonthYearTitle), 'dd MMMM, yyyy');
}

export enum DefaultTime {
  from = '09:00',
  to = '17:00',
}

export const CalendarWeekDays = [
  'CALENDAR_WEEK_DAYS.MONDAY',
  'CALENDAR_WEEK_DAYS.TUESDAY',
  'CALENDAR_WEEK_DAYS.WEDNESDAY',
  'CALENDAR_WEEK_DAYS.THURSDAY',
  'CALENDAR_WEEK_DAYS.FRIDAY',
  'CALENDAR_WEEK_DAYS.SATURDAY',
  'CALENDAR_WEEK_DAYS.SUNDAY',
];

export const ColStartClasses = [
  'col-start-7',
  '',
  'col-start-2',
  'col-start-3',
  'col-start-4',
  'col-start-5',
  'col-start-6',
];
export function getErrorMessage(arg: HttpErrorResponse): string {
  if (arg.error) {
    if (arg.error.data && typeof arg.error.data === 'object' && arg.error.data[Object.keys(arg.error.data)[0]]) {
      return arg.error.data[Object.keys(arg.error.data)[0]];
    }
    return arg.error.message;
  }

  return arg.message;
}
export function noWhitespaceValidator(control: AbstractControl): ValidationErrors | null {
  if (!control.value) {
    return null;
  }
  return control.value.trim().length ? null : { whitespace: true };
}
export const DEFAULT_ELEMENT_LENGTHS = {
  BOOKING_TEMPLATE_NAME_LENGTH: 40,
  BOOKING_TEMPLATE_COMMENTS_LENGTH: 200,
};

export function getDaysInMonthsData(daysInMonth: Date[], appointmentsData: AppointmentDetails[], startDay: Date) {
  const daysWithEventsData = daysInMonth.map(calendarDay => {
    const appointmentArray: EventData[] = [];

    appointmentsData.forEach(appointmentDay => {
      const appointmentDate = parse(appointmentDay.date, 'yyyy-MM-dd', new Date());
      if (appointmentDate.getTime() === calendarDay.getTime()) {
        appointmentArray.push({
          appointmentTime: `${appointmentDay.startTime}   ${appointmentDay.name}`,
          appointmentId: appointmentDay.id as number,
          appointmentType: appointmentDay.bookingMeetingType.id as number,
          appointmentStatus: appointmentDay.status,
        });
      }
    });
    return {
      date: calendarDay,
      eventsData: appointmentArray.length > 0 ? [...appointmentArray] : [],
      isSameMonth: isSameMonth(calendarDay, startDay),
    };
  });

  return daysWithEventsData;
}

export function guessDefaultTimezone(): string {
  const clientTimeZone = (Intl as any).DateTimeFormat().resolvedOptions().timeZone;
  if (MAPPED_TIMEZONES.hasOwnProperty(clientTimeZone)) {
    return MAPPED_TIMEZONES[clientTimeZone];
  } else {
    return clientTimeZone;
  }
}

export function getSelectedTab(tabItems: Tab[], tabValue: string) {
  return tabItems.map(tab => {
    if (tab.urlSegment === tabValue) {
      return {
        ...tab,
        isSelected: true,
      };
    } else {
      return {
        ...tab,
        isSelected: false,
      };
    }
  });
}

export function downloadFile(fileContent: Blob, fileTitle: string): void {
  const url = window.URL.createObjectURL(fileContent);
  const a = document.createElement('a');
  a.style.display = 'none';
  a.href = url;
  a.download = fileTitle;
  a.click();
  document.body.appendChild(a);
  window.URL.revokeObjectURL(url);
  a.remove();
}
export function buildSettingFormData(vendorSetting: SaveVendorSetting): FormData {
  const formData: FormData = new FormData();
  if (vendorSetting.logo) {
    formData.append('logo', vendorSetting.logo, vendorSetting.logo.name);
  }

  vendorSetting.meetingTypes.forEach((meetingTypeId, index) => {
    formData.append(`meetingTypes[${index}]`, `${meetingTypeId}`);
  });

  formData.append('phoneNumber', vendorSetting.phoneNumber);
  formData.append('primaryColor', vendorSetting.primaryColor);
  formData.append('secondaryColor', vendorSetting.secondaryColor);

  return formData;
}
export function showPagination(pageBase: PageBase): boolean {
  if (pageBase.total <= pageBase.limit) {
    return false;
  }
  return true;
}

export function isBeyondFirstOrLastPage(pageBase: PageBase, currentPage: number): boolean {
  if (currentPage < 1) {
    return true;
  }

  const pageSize: number = Math.ceil(pageBase.total / pageBase.limit);
  if (currentPage > pageSize) {
    return true;
  }

  return false;
}

export function getCookie(name: string): string | null {
  const cookieName = `${name}=`;
  const decodedCookie = decodeURIComponent(document.cookie);

  const foundCookie = decodedCookie
    .split(';')
    .map(cookie => cookie.trim())
    .find(cookie => cookie.startsWith(cookieName));

  return foundCookie ? foundCookie.substring(cookieName.length) : null;
}

export function isNoScheduleExist(availabilities: Availability[]): boolean {
  return availabilities.map(availability => availability.isAvailable).some(availability => availability === true);
}
export function saveIntegrationInfo(integrationInfo: GoogleIntegrationInfo) {
  const localStorageKey: LocalStorageKey = 'integrationInfo';
  localStorage.setItem(localStorageKey, JSON.stringify(integrationInfo));
}
export function getIntegrationInfo(): GoogleIntegrationInfo | null {
  const localStorageKey: LocalStorageKey = 'integrationInfo';
  const integrationInfo = localStorage.getItem(localStorageKey);
  if (integrationInfo) {
    return JSON.parse(integrationInfo);
  }
  return null;
}
export function saveBookingSnapshot(bookingSnapshot: BookingSnapshot) {
  const localStorageKey: LocalStorageKey = 'bookingSnapshot';
  localStorage.setItem(localStorageKey, JSON.stringify(bookingSnapshot));
}
export function getBookingSnapshot(): BookingSnapshot | null {
  const localStorageKey: LocalStorageKey = 'bookingSnapshot';
  const bookingSnapshot = localStorage.getItem(localStorageKey);
  localStorage.removeItem(localStorageKey);
  if (bookingSnapshot) {
    return JSON.parse(bookingSnapshot);
  }
  return null;
}
