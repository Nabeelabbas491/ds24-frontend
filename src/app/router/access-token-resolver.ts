import { ActivatedRouteSnapshot, ResolveFn, RouterStateSnapshot } from '@angular/router';
import { getCookie } from '../shared/common/util';

export const accessTokenResolver: ResolveFn<void> = (route: ActivatedRouteSnapshot, state: RouterStateSnapshot) => {
  const token = getCookie('access_token');

  if (!token && state && route && location.href.includes('access_token=')) {
    document.cookie = `access_token=${
      location.href.split('access_token=')[1]
    }; path=/; expires=Fri, 31 Dec 9999 23:59:59 GMT; SameSite=None; Secure; Expires=`;
  }
};
