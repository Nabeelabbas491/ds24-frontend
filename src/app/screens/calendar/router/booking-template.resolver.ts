import { inject } from '@angular/core';
import { ActivatedRouteSnapshot, ResolveFn } from '@angular/router';
import { Store } from '@ngrx/store';
import * as actions from './../../../store/actions';
import { getBookingSnapshot } from './../../../shared/common/util';
import { BookingSnapshot } from 'src/app/types/booking.type';

export const bookingTemplateResolver: ResolveFn<void> = (route: ActivatedRouteSnapshot) => {
  const store = inject(Store);
  const isSync = route.queryParamMap.get('isSync');
  if (isSync) {
    const bookingSnapshot: BookingSnapshot | null = getBookingSnapshot();
    if (bookingSnapshot) {
      store.dispatch(actions.booking.loadBookingSnapshot({ bookingSnapshot }));
      store.dispatch(actions.booking.isBookingSynced({ isBookingSynced: true }));
    }
  } else {
    store.dispatch(
      actions.bookingAPI.getBookingTemplate({
        paramDetails: { paramId: route.paramMap.get('pid') as string, isClient: true },
      }),
    );
    store.dispatch(actions.timeZoneAPI.getTimeZones());
  }
};
