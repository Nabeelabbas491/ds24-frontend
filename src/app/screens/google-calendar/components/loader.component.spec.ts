import { ComponentFixture, TestBed } from '@angular/core/testing';

import { LoaderComponent } from './loader.component';
import { MockStore, provideMockStore } from '@ngrx/store/testing';
import { State, actions } from './../../../store';
import { RouterTestingModule } from '@angular/router/testing';
import { ActivatedRoute, convertToParamMap } from '@angular/router';
import { GoogleIntegrationInfo } from '../../../types/google-integration.type';
import { saveIntegrationInfo } from '../../../shared/common/util';

describe('LoaderComponent', () => {
  let component: LoaderComponent;
  let fixture: ComponentFixture<LoaderComponent>;
  let mockStore: MockStore<State>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [LoaderComponent],
      imports: [RouterTestingModule],
      providers: [
        provideMockStore(),
        {
          provide: ActivatedRoute,
          useValue: {
            snapshot: {
              queryParamMap: convertToParamMap({ code: 'xyz' }),
            },
          },
        },
      ],
    });
    fixture = TestBed.createComponent(LoaderComponent);
    component = fixture.componentInstance;
    mockStore = TestBed.inject(MockStore);
    fixture.detectChanges();
    jest.spyOn(mockStore, 'dispatch');
  });

  it('should compile', () => {
    fixture.detectChanges();
    expect(component).toMatchSnapshot();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should call the syncAppointmentClient if the integrationMode is client-appointment', () => {
    const code = 'xyz';
    const integrationInfo: GoogleIntegrationInfo = {
      integrationMode: 'client-booking',
      bookingProductId: '123',
    };

    saveIntegrationInfo(integrationInfo);

    component.ngOnInit();
    fixture.detectChanges();

    expect(mockStore.dispatch).toHaveBeenCalledWith(
      actions.googleIntegration.syncAppointmentClient({
        bookingProductId: integrationInfo.bookingProductId as string,
        code,
      }),
    );
  });

  it('should call the googleConnect if the integrationMode is vendor-appointment', () => {
    const code = 'xyz';
    const integrationInfo: GoogleIntegrationInfo = {
      integrationMode: 'vendor-appointment',
      appointmentId: '123',
    };

    saveIntegrationInfo(integrationInfo);

    component.ngOnInit();
    fixture.detectChanges();

    expect(mockStore.dispatch).toHaveBeenCalledWith(
      actions.googleIntegration.googleConnect({
        code,
      }),
    );
  });
});
