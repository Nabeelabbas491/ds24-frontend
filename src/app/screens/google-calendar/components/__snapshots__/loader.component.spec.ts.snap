// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoaderComponent should compile 1`] = `
LoaderComponent {
  "__ngContext__": 0,
  "activatedRoute": {
    "snapshot": {
      "queryParamMap": ParamsAsMap {
        "params": {
          "code": "xyz",
        },
      },
    },
  },
  "spinnerColor": "fill-neutral-n500",
  "store": _MockStore {
    "actionsObserver": _ActionsSubject {
      "_value": {
        "type": "@ngrx/store/init",
      },
      "closed": false,
      "currentObservers": null,
      "hasError": false,
      "isStopped": false,
      "observers": [],
      "thrownError": null,
    },
    "dispatch": [MockFunction],
    "initialState": {},
    "lastState": {},
    "reducerManager": _MockReducerManager {
      "_value": [Function],
      "closed": false,
      "currentObservers": null,
      "hasError": false,
      "isStopped": false,
      "observers": [],
      "thrownError": null,
    },
    "scannedActions$": Observable {
      "source": _ActionsSubject {
        "_value": {
          "type": "@ngrx/store/init",
        },
        "closed": false,
        "currentObservers": null,
        "hasError": false,
        "isStopped": false,
        "observers": [],
        "thrownError": null,
      },
    },
    "selectors": Map {},
    "source": _MockState {
      "_value": {},
      "closed": false,
      "currentObservers": [
        SafeSubscriber {
          "_finalizers": [
            Subscription {
              "_finalizers": null,
              "_parentage": [Circular],
              "closed": false,
              "initialTeardown": [Function],
            },
          ],
          "_parentage": null,
          "closed": false,
          "destination": ConsumerObserver {
            "partialObserver": {
              "error": [Function],
              "next": [Function],
            },
          },
          "initialTeardown": undefined,
          "isStopped": false,
        },
      ],
      "hasError": false,
      "isStopped": false,
      "observers": [
        SafeSubscriber {
          "_finalizers": [
            Subscription {
              "_finalizers": null,
              "_parentage": [Circular],
              "closed": false,
              "initialTeardown": [Function],
            },
          ],
          "_parentage": null,
          "closed": false,
          "destination": ConsumerObserver {
            "partialObserver": {
              "error": [Function],
              "next": [Function],
            },
          },
          "initialTeardown": undefined,
          "isStopped": false,
        },
      ],
      "state": [Function],
      "thrownError": null,
    },
    "state": [Function],
    "state$": _MockState {
      "_value": {},
      "closed": false,
      "currentObservers": [
        SafeSubscriber {
          "_finalizers": [
            Subscription {
              "_finalizers": null,
              "_parentage": [Circular],
              "closed": false,
              "initialTeardown": [Function],
            },
          ],
          "_parentage": null,
          "closed": false,
          "destination": ConsumerObserver {
            "partialObserver": {
              "error": [Function],
              "next": [Function],
            },
          },
          "initialTeardown": undefined,
          "isStopped": false,
        },
      ],
      "hasError": false,
      "isStopped": false,
      "observers": [
        SafeSubscriber {
          "_finalizers": [
            Subscription {
              "_finalizers": null,
              "_parentage": [Circular],
              "closed": false,
              "initialTeardown": [Function],
            },
          ],
          "_parentage": null,
          "closed": false,
          "destination": ConsumerObserver {
            "partialObserver": {
              "error": [Function],
              "next": [Function],
            },
          },
          "initialTeardown": undefined,
          "isStopped": false,
        },
      ],
      "state": [Function],
      "thrownError": null,
    },
  },
}
`;
